File: 0. Easy/0669. Trim a Binary Search Tree/trim_tree.py
# Prompt: https://leetcode.com/problems/trim-a-binary-search-tree

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

# Runtime: 60 ms, faster than 77.33% of Python3 online submissions for Trim a Binary Search Tree.
# Memory Usage: 17.4 MB, less than 5.36% of Python3 online submissions for Trim a Binary Search Tree.

class Solution:
    def findNodesWithinRange(self, node: TreeNode, L: int, R: int, toKeep: List[int]) -> None:
        if node != None:
            if node.val >= L and node.val <= R:
                toKeep.append(node.val)
                self.findNodesWithinRange(node.left, L, R, toKeep)
                self.findNodesWithinRange(node.right, L, R, toKeep)
            elif node.val > R:
                self.findNodesWithinRange(node.left, L, R, toKeep)
            elif node.val < L:
                self.findNodesWithinRange(node.right, L, R, toKeep)
                
    def addNode(self, node: TreeNode, val: int):
        if node.val > val:
            if node.left == None:
                node.left = TreeNode(val)
            else:
                self.addNode(node.left, val)
        else:
            if node.right == None:
                node.right = TreeNode(val)
            else:
                self.addNode(node.right, val)
            
    def trimBST(self, root: TreeNode, L: int, R: int) -> TreeNode:
        # to keep
        toKeep = []
        self.findNodesWithinRange(root, L, R, toKeep)
        if len(toKeep) > 0:
            i = 0
            newTree = TreeNode(toKeep[i])
            i += 1
            while i < len(toKeep):
                self.addNode(newTree, toKeep[i])
                i += 1
            return newTree
        return []

File: 0. Easy/0669. Trim a Binary Search Tree/trim_tree.cpp
// Prompt: https://leetcode.com/problems/trim-a-binary-search-tree
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */

// Runtime: 28 ms, faster than 43.10% of C++ online submissions for Trim a Binary Search Tree.
// Memory Usage: 22.1 MB, less than 23.66% of C++ online submissions for Trim a Binary Search Tree.
#include <queue>
class Solution {
public:
    void findNodesWithinRange(TreeNode* node, int L, int R, std::queue<int>& keep) {
        if (node != NULL) {
            if (node->val <= R && node->val >= L) {
                keep.push(node->val);
                findNodesWithinRange(node->left, L, R, keep);
                findNodesWithinRange(node->right, L, R, keep);
            } else if (node->val > R) {
                findNodesWithinRange(node->left, L, R, keep);
            } else if (node->val < L) {
                findNodesWithinRange(node->right, L, R, keep);
            }
        }
    }
    void addNode(TreeNode* root, int val) {
        if (root->val < val) {
            if (root->right == NULL) {
                root->right = new TreeNode(val);
            } else {
                addNode(root->right, val);                
            }
        } else {
            if (root->left == NULL) {
                root->left = new TreeNode(val);
            } else {
                addNode(root->left, val);                
            }
        }
    }
    
    TreeNode* trimBST(TreeNode* root, int L, int R) {
        std::queue<int> nodesToKeep;
        findNodesWithinRange(root, L, R, nodesToKeep);
        // add them to a tree
        if (!nodesToKeep.empty()) {
            TreeNode* newTree = new TreeNode(nodesToKeep.front());
            nodesToKeep.pop();
            while (!nodesToKeep.empty()) {
                addNode(newTree, nodesToKeep.front());
                nodesToKeep.pop();
            }
            return newTree;
        } else {
            return NULL;
        }
    }
};
